New layers

spencerduncan opened the pull request:
- Added max shields and resources as unit data. Resources was separated from HP so that resources of Refineries and such can be seen.
- Added minerals, gas, energy and shields to get_layer and visual example.lua
- Added function is_gas_geyser to utils.lua
- Added functions get_feature and draw_values to init.lua
    - get_feature is similar to get_layer, but outputs a 2D tensor instead of a 3D RGB tensor
    - get_feature does not preform any scaling on values unlike get_layer, besides adding 1 to certain 0 indexed values
    - draw_values is get_features version of draw_entities
    - get_feature can handle any member variable of units, although some tweaking may be required for untested types
    - added test script tensor_example.lua to examples. All cases used in this script have been tested.
- Fixed a bug in get_layer where a unit being in the top-left corner of the screen could cause the script to crash
- Changed default skip_frames in visual_example.lua to 30 to save disk space.
- Added option --micro_game to visual_example.lua for testing of resource layers (mins/gas)
- Added compiled BWEnv.exe and BWEnv.dll

spencerduncan committed:
Merge pull request #1 from TorchCraft/master

Nothing interesting.

spencerduncan committed:
First pass at new layer types. Added minerals, energy, and shields. Gas is NYI, pending BWEnv changes.

spencerduncan committed:
Added resources separate from HP to frame; allowing for gas to be read from
	Refineries etc

Added max_shields to frame;

Added function get_feature to init.lua for making sparse tensors of feature data

Added function is_gas_geyser to utils.lua

spencerduncan committed:
Fixed some typos

spencerduncan committed:
Added new builds off BWEnv to support new data types

Changed .gitmodules to use HTTPS rather that git://

spencerduncan committed:
Added max_shield and resourcesw to frame.cpp

spencerduncan committed:
Fixed a typo in frame.cpp (colon instead of semicolon :/)

Recompiled BWEnv with previous commit

spencerduncan committed:
Maybe fixed frame_lua.cpp

spencerduncan committed:
Last couple of silly fixes to draw layer; It actually works now

spencerduncan committed:
Tweaked visual example so that minerals and gas are not drawn in micro_battle mode

Fixed issue where script would crash when drawing units near the top left corner
	of the FoV

spencerduncan committed:
Added usage example / test environment for get_feature
Fixed issues with new features in init.lua

edran reviewed (changes requested):
Reviewing this soon :)

edran commented on .gitmodules:
```diff
@@ -1,6 +1,6 @@
 [submodule "BWEnv/include/czmq"]
 	path = BWEnv/include/czmq
-	url = git@github.com:torchcraft/czmq
+	url = https://github.com/torchcraft/czmq
```
Please make "torchcraft" consistent. Either lowercase or CamelCase.

spencerduncan replied:
I'm think it's consistently lowercase everywhere but the repo name.

edran commented on init.lua:
```diff
@@ -1333,20 +1338,162 @@ function torchcraft:get_layer(layer_type, also_enemy)
                                        32, 32, color)
             end
         end
+    elseif layer_type == "minerals" then
```
Nice! I'd rather have those layers merged though (possibly showing blue for minerals and green for grass).

spencerduncan replied:
Yeah, either make it green or keep it greyscale like in get_feature and hope the nets can infer the difference from the Type layer. I might write another function that takes a list of features and returns a single merged tensor.

edran commented on init.lua:
```diff
@@ -1333,20 +1338,162 @@ function torchcraft:get_layer(layer_type, also_enemy)
+--        Or just render them to colors with draw value to increase spacing
+
+    local img
+    if dtype == "short" then
```
In Torch you can easily create different tensor types, so there is no need to have an additional argument specifying this. Also `img` is instantiated earlier and should be reused.

spencerduncan replied:
Although, I'm not seeing img being instantiated anywhere that isn't as a local inside another function.

spencerduncan replied:
Could you push that so I can make my code work with it?

edran replied:
I'll do the merge myself after the PR :) for now leave `local img`.

edran commented on init.lua:
```diff
@@ -1333,20 +1338,162 @@ function torchcraft:get_layer(layer_type, also_enemy)
+                                       32, 32, value)
+            end
+        end
+        return img
```
Too many branches in this method - earlier the logic had a few `elseifs` and one return (or something like) that, which is preferable for debugging.

spencerduncan replied:
I cleaned this up some. I'll probably change it some more once I see how you're handling the img variable.

edran commented on init.lua:
```diff
@@ -1333,20 +1338,162 @@ function torchcraft:get_layer(layer_type, also_enemy)
+        end
+        return img
+    elseif feature == "gas" then
+        t = tablex.merge(self.state.units_myself, self.state.units_neutral,true)
```
s/`,true`/`, true`/

spencerduncan replied:
Fixed.

edran commented on BWEnv/src/controller.cc:
```diff
@@ -782,17 +782,11 @@ void Controller::addUnit(BWAPI::Unit u, replayer::Frame& frame, BWAPI::PlayerInt
     + utype.airWeapon().damageBonus()
     * player->getUpgradeLevel(utype.airWeapon().upgradeType()))
     * utype.maxAirHits() * utype.airWeapon().damageFactor();
-  int max_health = utype.maxHitPoints();
```
Are you sure there are no side effects with this change?

(cc: @syhw )

spencerduncan replied:
max_health wasn't used anywhere besides being passed to the unit constructor. BWAPI::UnitType::maxHitPoints already returns an int, so it can't be a casting issue. Now it just has maxHitPoints being passed directly to the constructor.

Calling getHitPoints() on minerals / geysers should return 1, so that shouldn't cause any issues.

edran reviewed (changes requested):
Thank you for the PR! A few things to change / check.

Also please remove the binaries from the branch - history included - as we don't want to pollute the repository with them (we use GitHub's release mechanism to distribute them).

syhw reviewed (commented):
please do not hard-code, e.g. use `ut.Resource_Mineral_Field` instead of 176.

ebetica commented on replayer/frame.cpp:
```diff
@@ -12,7 +12,7 @@
   out << o.id << " " << o.x << " " << o.y << " "
     << o.health << " " << o.max_health << " "
-    << o.shield << " " << o.energy << " " << o.maxCD << " "
+    << o.shield << " " << o.max_shield << " " << o.energy << " " << o.maxCD << " "
```
Please update the torchcraft protocol version

spencerduncan committed:
Removing BWEnv binaries

spencerduncan committed:
Changes per PR review

Updated protocol version in ZMQ_server.h and init.lua

Moved is_gas_geyser and is_mineral_field to init.lua. These functions now
work like is_worker and take a unit type rather than a unit.

Cleaned up get_feature some. Now defaults to ShortTensor, and doesn't have
datatype as an argument. Now only has one return statement. Updated
tensor_example to match these changes.

Added a missing space to init.lua.

These changes are currently untested; I'll test when I have access to my
Linux box and post an update in the PR thread.

spencerduncan committed:
Fixed some typos. New commit has been tested with HP layer.

spencerduncan commented:
This last commit should fix the issues people bought up.
